kosat-core/src/commonMain/kotlin/org/kosat/CDCL.kt

```diff
@@ -405,6 +415,136 @@ class CDCL {
+        if (trail.size == numberOfVariables) {
+            return getModel()
+        }
+
+        return null
+    }
+
+    /**
+     * Literal probing is only useful if the probe can lead
+     * lead to derivation of something by itself. We can
+     * generate list of possibly useful probes ahead of time.
+     *
+     * TODO: Right now, this implementation is very naive.
+     *       If there is a cycle in binary implication graph,
+     *       we will generate a lot of useless probes that will
+     *       propagate in exactly the same way and possibly
+     *       produce a lot of duplicate/implied binary clauses.
+     *       This will be fixed with the implementation of
+     *       Equivalent Literal Substitution (ELS).
+     *
+     *  @return list of literals to try probing with
+     */
+    private fun generateProbes(): List<Lit> {
+        val probes = mutableSetOf<Lit>()
+
+        for (clause in constraints) {
+            // (A | B) <==> (-A -> B) <==> (-B -> A)
+            // Both -A and -B can be used as probes
+            if (clause.size == 2) {
+                val (a, b) = clause
+                if (getValue(a) == LBool.UNDEFINED) probes.add(a.neg)
+                if (probes.size >= flpMaxProbes) break
+                if (getValue(b) == LBool.UNDEFINED) probes.add(b.neg)
+                if (probes.size >= flpMaxProbes) break
+            }
+        }
+
+        return probes.toMutableList()
+    }
+
+    /**
+     * Try to propagate each probe and see if it leads to
+     * deduction of new binary clauses, or maybe even a
+     * conflict.
+     *
+     * Consider clauses (-1, 2), (-1, -2, 3), and a probe 1.
+     * By propagating 1, we can derive 2, which in turn
+     * allows us to derive 3. This means that we can add
+     * a new binary clause (-1, 3) to the problem. There are
+     * other mechanisms that can derive this clause, but
+     * probing is one of them.
+     */
+    private fun failedLiteralProbing(): Model? {
+        val probesToTry = generateProbes()
+
+        for (probe in probesToTry) {
+            // If we know that the probe is already true, just skip it.
+            if (getValue(probe) == LBool.TRUE) {
+                continue
+            }
+
+            level++
+            uncheckedEnqueue(probe)
+            val startOfProbeTrail = trail.size
+            val conflictClause = propagate()
+
+            if (conflictClause != null) {
+                // Assigning a value to a probe lead to a conflict
+                // immediately means that the probe must be false.
+                clearTrail(0)
+                level = 0
+                qhead = trail.size
+
+                if (getValue(probe) == LBool.TRUE) {
+                    return Model.UNSAT
+                }
+
+                uncheckedEnqueue(probe.neg)
+
+                // Can we learn more while we are at level 0?
+                propagate()?.let { return Model.UNSAT }
```

Comment:
Consider NOT using `let`. This Kotlin feature is not easy to understand for non-Kotlin devs.